src/base/libmesh.C

```diff
@@ -798,7 +816,20 @@ LibMeshInit::~LibMeshInit()
 #endif
 
       if (libmesh_initialized_mpi)
-        MPI_Finalize();
+        {
+          // We can't just libmesh_assert here because destructor,
+          // but we ought to report any errors
+          unsigned int error_code = MPI_Finalize();
```

Comment:
I honestly can't understand how *anything* sane could explain the different behavior I observed.  I mean, I guess I could write an MPI implementation that forks and nohups a trojan which then kills its parent shell a fraction of a second after exit, but how does the parent shell die in between two echo commands by *accident*?

MPI_Finalize() is only valid to call after all non-blocking communication has been resolved, so I can see how we might have a bug earlier in a program which would only trigger at the final cleanup... but what I'm seeing gets *past* the final cleanup and even past program exit.